File: tests-playwright/src/pages/dynamic/dynamic-page.ts
import { expect, type Locator, type Page } from '@playwright/test'
import { routes } from '../../constants/routes'
import { BasePage } from '../base-page'

/** Page object for the dynamic elements page: ticker, delayed button, lazy list. */
export class DynamicPage extends BasePage {
  readonly root: Locator
  readonly tickerCard: Locator
  readonly delayedButton: Locator
  readonly lazyList: Locator

  constructor(page: Page) {
    super(page)
    this.root = page.getByTestId('dynamic-page')
    this.tickerCard = page.getByTestId('ticker-card')
    this.delayedButton = page.getByTestId('delayed-button')
    this.lazyList = page.getByTestId('lazy-list')
  }

  async open(): Promise<void> {
    await this.goto(routes.dynamic)
    await expect(this.root).toBeVisible()
  }

  async verifyShellReady(): Promise<void> {
    await expect(this.tickerCard).toBeVisible()
    await expect(this.delayedButton).toBeVisible()
    await expect(this.lazyList).toBeVisible()
  }
}

File: tests-selenium/src/test/java/com/coinpulse/selenium/pages/DynamicPage.java
package com.coinpulse.selenium.pages;

import com.coinpulse.selenium.constants.Routes;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

/**
 * Dynamic elements page object; selectors mirror the Cypress page object.
 */
public class DynamicPage extends BasePage {

    private static final By PAGE = byTestId("dynamic-page");
    private static final By TICKER_CARD = byTestId("ticker-card");
    private static final By DELAYED_BUTTON = byTestId("delayed-button");
    private static final By LAZY_LIST = byTestId("lazy-list");

    public DynamicPage(WebDriver driver) {
        super(driver);
    }

    public void open() {
        openPath(Routes.DYNAMIC);
        awaitVisible(PAGE);
    }

    public void verifyShellReady() {
        awaitVisible(TICKER_CARD);
        awaitVisible(DELAYED_BUTTON);
        awaitVisible(LAZY_LIST);
    }
}
